flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/ExpireSnapshotsProcessor.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Calls the {@link ExpireSnapshots} to remove the old snapshots and emits the filenames which could
+ * be removed in the {@link #DELETE_STREAM} side output.
+ */
+public class ExpireSnapshotsProcessor extends ProcessFunction<Trigger, TaskResult> {
+  private static final Logger LOG = LoggerFactory.getLogger(ExpireSnapshotsProcessor.class);
+  public static final OutputTag<String> DELETE_STREAM =
+      new OutputTag<>("expire-snapshots-file-deletes-stream", Types.STRING);
+
+  private final TableLoader tableLoader;
+  private final Long maxSnapshotAgeMs;
+  private final Integer retainLast;
+  private final int plannerPoolSize;
+  private transient ExecutorService plannerPool;
+  private transient Table table;
+
+  public ExpireSnapshotsProcessor(
+      TableLoader tableLoader, Long maxSnapshotAgeMs, Integer retainLast, int plannerPoolSize) {
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+
+    this.tableLoader = tableLoader;
+    this.maxSnapshotAgeMs = maxSnapshotAgeMs;
+    this.retainLast = retainLast;
+    this.plannerPoolSize = plannerPoolSize;
+  }
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    tableLoader.open();
+    this.table = tableLoader.loadTable();
+    this.plannerPool = ThreadPools.newWorkerPool(table.name() + "-table--planner", plannerPoolSize);
+  }
+
+  @Override
+  public void processElement(Trigger trigger, Context ctx, Collector<TaskResult> out)
+      throws Exception {
+    try {
+      table.refresh();
+      ExpireSnapshots expireSnapshots = table.expireSnapshots();
+      if (maxSnapshotAgeMs != null) {
+        expireSnapshots = expireSnapshots.expireOlderThan(ctx.timestamp() - maxSnapshotAgeMs);
+      }
+
+      if (retainLast != null) {
+        expireSnapshots = expireSnapshots.retainLast(retainLast);
+      }
+
+      AtomicLong deleteFileCounter = new AtomicLong(0L);
+      expireSnapshots
+          .planWith(plannerPool)
+          .deleteWith(
+              file -> {
+                ctx.output(DELETE_STREAM, file);
+                deleteFileCounter.incrementAndGet();
+              })
+          .cleanExpiredFiles(true)
+          .commit();
+
+      LOG.info(
+          "Successfully finished expiring snapshots for {} at {}. Scheduled {} files for delete.",
+          table,
+          ctx.timestamp(),
+          deleteFileCounter.get());
+      out.collect(
+          new TaskResult(trigger.taskId(), trigger.timestamp(), true, Collections.emptyList()));
+    } catch (Exception e) {
+      LOG.info("Exception expiring snapshots for {} at {}", table, ctx.timestamp(), e);
+      out.collect(
+          new TaskResult(trigger.taskId(), trigger.timestamp(), false, Lists.newArrayList(e)));
```

Comment:
Other maintenance tasks might have multiple errors from multiple operators/subtasks

Reply:
hmm. I am still not quite following. Each operator subtask emits a `TaskResult`. Each `TaskResult` should only contain one exception, right?

I didn't see the exceptions are used by downstream. if `success` boolean flag good enough for downstream, maybe we can remove the exceptions from `TaskResult` as stack trace can be non-trivial.

BTW, `TaskResult` is not marked as `Serializable`.

Reply:
When we do compaction, then we have multiple subtasks running parallel. If more than one of them fails, then we will have multiple exception messages to report, but we don't want to fail the job (especially in PostCommitTopology).

Reply:
Also I think it is very rare that we have an exception, and it is good to have a single place where we can collect/handle those. So while serializing a stack trace is non-trivial, I think it worth the cost in the long run.

Fixed the TaskResult serialization....

Reply:
> When we do compaction, then we have multiple subtasks running parallel. If more than one of them fails, then we will have multiple exception messages to report

I thought `TaskResult` is created and emitted by every subtask. Do we have `TaskResult` to merge multiple `TaskResult` objects in the downstream operator?

Reply:
Sorry, I was not clear enough.
The `TaskResult` is only the result of the MaintenanceTasks.
The `MaintenanceTask` can have multiple operators, and these operators could fail independently.
For example for the Orphan file removal:
- We can have an error listing the FileSystem files (this could happen after some of the file names are already emitted)
- We can have an error listing reading the Table metadata (this could happen after some of the files are already emitted)

This means that for the end result of the tasks we will have multiple exceptions, which I think we should report.

Reply:
Let me make sure I understand. Orphan file maintenance task can have multiple operators (like listing file from storage, retrieve live files from table). `TaskResult` can propagate through the sub DAG and combine exceptions if there are.

Reply:
These operators don't emit `TaskResult` objects. They are internal details for the orphan files removal flow. The flow collects the exceptions and emits the `TaskResult` based on those.